File: lamb-upeu-sis/src/main/java/unmsm/edu/pe/tesis/domain/services/SecretariaDefensaService.java
package unmsm.edu.pe.tesis.domain.services;

import unmsm.edu.pe.shared.response.PageResponse;
import unmsm.edu.pe.tesis.application.dto.ExpedienteBandejaItem;
import unmsm.edu.pe.tesis.application.dto.RubricaBandejaItem;

import java.util.List;
import java.util.UUID;

/** Secretaría · Etapa 5 (Defensa) — paso 1: recibir el expediente y comunicar al Coordinador. */
public interface SecretariaDefensaService {

    PageResponse<ExpedienteBandejaItem> bandeja(String buscar, int page, int size);

    /** La Secretaría recibe el expediente y lo comunica al Coordinador del Programa. */
    void recibir(UUID tesisId);

    /** Proyectos con revisores designados que requieren (o ya tienen) la rúbrica oficial subida. */
    List<RubricaBandejaItem> bandejaRubricas(String buscar);

    /** La Secretaría sube la rúbrica oficial (Word .docx) para habilitar la evaluación de los revisores. */
    /** Enciende/apaga la evaluación con la rúbrica oficial del sistema (reemplaza a subir el Word). */
    void habilitarRubrica(UUID tesisId, boolean habilitar);

    /** Legado: adjuntar un Word propio del expediente. Se mantiene para casos excepcionales. */
    void subirRubrica(UUID tesisId, byte[] contenido, String nombreOriginal, String contentType);

    /** Vista previa (texto) de la rúbrica subida, para confirmar que es el documento correcto (no descarga). */
    List<String> previsualizarRubrica(UUID tesisId);

    /** El Word (.docx) crudo de la rúbrica, para renderizarlo en el navegador (vista previa fiel). */
    unmsm.edu.pe.tesis.application.dto.ArchivoDescargable documentoRubrica(UUID tesisId);

    // ── Programación de la defensa (la realiza la Secretaría) ──

    /** Info de la defensa: si está programada, fecha/hora/lugar, dictamen y jurado. */
    unmsm.edu.pe.tesis.application.dto.DefensaInfo defensa(UUID tesisId);

    /** Docentes de la línea de la tesis, seleccionables como Jurado Examinador. */
    List<unmsm.edu.pe.tesis.application.dto.DocenteOpcion> docentesDefensa(UUID tesisId);

    /** Programa la defensa: Jurado Examinador (presidente + 2 miembros + asesor) + fecha/hora/lugar. */
    void programarDefensa(UUID tesisId, unmsm.edu.pe.tesis.application.dto.ProgramarDefensaRequest req);
}

File: themeforest-Fj3VzBCp-fuse-angularjs-material-design-admin-template/fuse/src/app/views/dashboard/secretaria-defensa/services/secretaria-defensa.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '@/environments/environment';
import { END_POINTS } from '@/app/providers/utils/end-points';

@Injectable({ providedIn: 'root' })
export class SecretariaDefensaService {
  private _http = inject(HttpClient);
  private readonly base = environment.url + END_POINTS.secretaria.defensa;

  bandeja$(buscar?: string, page = 0, size = 20): Observable<any> {
    let params = new HttpParams().set('page', String(page)).set('size', String(size));
    if (buscar) params = params.set('buscar', buscar);
    return this._http.get<any>(this.base, { params });
  }

  recibir$(tesisId: string): Observable<any> {
    return this._http.post<any>(`${this.base}/${tesisId}/recibir`, {});
  }

  /** Proyectos con revisores designados que requieren (o ya tienen) la rúbrica oficial. */
  bandejaRubricas$(buscar?: string): Observable<any> {
    let params = new HttpParams();
    if (buscar) params = params.set('buscar', buscar);
    return this._http.get<any>(`${this.base}/rubricas`, { params });
  }

  /** Enciende/suspende la evaluación con la rúbrica oficial del sistema (no sube nada). */
  habilitarRubrica$(tesisId: string, habilitar: boolean): Observable<any> {
    return this._http.post<any>(`${this.base}/${tesisId}/rubrica/habilitar?habilitar=${habilitar}`, {});
  }

  /** Word de la rúbrica oficial vigente del enfoque (para la vista previa). */
  rubricaOficialRaw$(enfoque: string): Observable<Blob> {
    return this._http.get(`${environment.url}api/secretaria/rubricas-oficiales/vigente/${enfoque}/documento`,
      { responseType: 'blob' });
  }

  /** Legado: adjunta un Word propio del expediente (caso excepcional). */
  subirRubrica$(tesisId: string, archivo: File): Observable<any> {
    const fd = new FormData();
    fd.append('archivo', archivo);
    return this._http.post<any>(`${this.base}/${tesisId}/rubrica`, fd);
  }

  /** Vista previa (texto) de la rúbrica subida, para confirmar que es el documento correcto. */
  previewRubrica$(tesisId: string): Observable<any> {
    return this._http.get<any>(`${this.base}/${tesisId}/rubrica/preview`);
  }

  /** El Word (.docx) crudo de la rúbrica, para renderizarlo en el navegador (vista previa fiel). */
  descargarRubricaRaw$(tesisId: string): Observable<Blob> {
    return this._http.get(`${this.base}/${tesisId}/rubrica/raw`, { responseType: 'blob' });
  }

  // ── Programación de la defensa (la realiza la Secretaría) ──
  defensa$(tesisId: string): Observable<any> {
    return this._http.get<any>(`${this.base}/${tesisId}/defensa`);
  }
  programarDefensa$(tesisId: string, body: any): Observable<any> {
    return this._http.post<any>(`${this.base}/${tesisId}/defensa`, body);
  }
}
